interp/notation.mli

```diff
@@ -323,6 +325,48 @@ val declare_custom_entry_has_ident : string -> int -> unit
 val entry_has_global : notation_entry_level -> bool
 val entry_has_ident : notation_entry_level -> bool
 
+(** Entry keys for constr notations *)
+
+type side = Left | Right
+
+type production_position =
+  | BorderProd of side * Gramlib.Gramext.g_assoc option
+  | InternalProd
+
+type production_level =
+  | NextLevel
+  | NumLevel of int
+  | DefaultLevel (** Interpreted differently at the border or inside a rule *)
+
+(** User-level types used to tell how to parse or interpret of the non-terminal *)
+
+type 'a constr_entry_key_gen =
+  | ETIdent
+  | ETGlobal
+  | ETBigint
+  | ETBinder of bool  (* open list of binders if true, closed list of binders otherwise *)
+  | ETConstr of Constrexpr.notation_entry * Notation_term.constr_as_binder_kind option * 'a
+  | ETPattern of bool * int option (* true = strict pattern, i.e. not a single variable *)
+
+(** Entries level (left-hand side of grammar rules) *)
+
+type constr_entry_key =
+    (production_level * production_position) constr_entry_key_gen
+
+(** Dealing with precedences *)
+
+type level = notation_entry * entry_level * entry_relative_level list * constr_entry_key list
```

Comment:
AFAICT interp only needs the `Constrexpr.entry_level` part, so maybe the other data should be left in the upper levels such that interp doesn't depend on gramlib.

Reply:
Ok, but the data is really the whole thing. This single integer taken out of other level data does not make much sense IMHO. What's the pb with interp depending on gramlib?

Reply:
Indeed, as @SkySkimmer pointed out interp doesn't depend on parsing. I guess not a huge problem having `interp` depend on the parsing engine but the idea was to enforce a more orthogonal codebase at that point. In particular, we were aiming for a self-contained parsing module, which is only depended by the toplevel / document managers. So that:

- pretyping: containing the glob_constr elaborated forms, + `pretype : glob_constr -> constr`
- interp: containing the const_expr AST + `interp: constr_expr -> glob_constr`
- parsing: a separate component + `parse : input -> constr_expr`

Indeed how to organize things with custom notations etc... is a bit tricky.

Reply:
The current separation enforces statically that parsing cannot be called from interpretation tho. Whether that's a useful property, I'm not sure, I'd say yes.

Reply:
I put the separation back.